graphql-dgs-spring-boot-micrometer/src/main/kotlin/com/netflix/graphql/dgs/metrics/micrometer/DgsGraphQLMetricsInstrumentation.kt

```diff
@@ -382,4 +446,13 @@ class DgsGraphQLMetricsInstrumentation(
 
         internal data class ErrorTagValues(val path: String, val type: String, val detail: String)
     }
+
+    internal object MeasurementUtils {
+        inline fun <reified T> serializeAndMeasureSize(obj: T): Int {
+            return ByteArrayOutputStream().use { outputStream ->
+                mapper.writeValue(outputStream, obj)
```

Comment:
This serializes the whole thing _just_ for the metric. Since JSON serialization is quite expensive, I'm expecting this to impact performance.

Reply:
Perhaps capturing the size of the request and response body in a ServletFilter might save on the overhead, since it's serialized already?

Reply:
I think a ServletFilter would be a more appropriate place. There's also some complexity because you can only read the body once in a filter. I'm not completely sure what the best way to get the size there is.

Either way, something like that needs to be done so that body is serialized/deserialized more than once. Extra serialization/deserialization won't be acceptable from a CPU perspective.

Reply:
Thank you for the feedback, I will investigate the ServletFilter approach

Reply:
After digging for a little bit, Spring has a `ContentCachingRequestWrapper` and `ContentCachingResponseWrapper` that takes care of the "can only read/write once" issue.

A filter like this seems to work: 

```

@Component
public class SizeFilter implements Filter {
    @Override
    public void init(FilterConfig filterConfig) throws ServletException {
        Filter.super.init(filterConfig);
    }

    @Override
    public void doFilter(ServletRequest request, ServletResponse response, FilterChain chain) throws IOException, ServletException {

        if (request instanceof HttpServletRequest servletRequest) {
            var wrappedRequest = new ContentCachingRequestWrapper(servletRequest);
            var wrappedResponse = new ContentCachingResponseWrapper((HttpServletResponse) response);
            System.out.println("Request content length: " + wrappedRequest.getContentLength());
            chain.doFilter(wrappedRequest, wrappedResponse);

            System.out.println("Response content length: " + wrappedResponse.getContentSize());
            wrappedResponse.copyBodyToResponse();

        } else {
            chain.doFilter(request, response);
        }
    }

    @Override
    public void destroy() {
        Filter.super.destroy();
    }
}
```

There will still be some overhead because it needs to copy bytes around, but it will be way less expensive than JSON (de)serialization.

This sits at the request level, so it's not GraphQL specific at all. It probably doesn't need to be?

Reply:
Another idea - just take the size of the query string as the size? It's not json at that point, but does that matter at all? Not sure how this metric is going to be used, but I expect it's to get an idea of the size of the query, not necessarily the exact JSON size

Reply:
Oh, if the query is available as a string, then that's all we need! assuming variables are also taken care of.

Reply:
Yeah thinking about it more, I don't think it's needed to serialize it, I thought serializing it would more accurately reflect the size but it might not be worth the tradeoff plus yeah, the metric is likely needed more to get an idea of the size, not the exact size and I'll double check with the user who requested the metric

For the variables, in the approach where we measure the size in the instrumentation code, the type of that and other components types: 
- variables (`Map<String, Object>`)
- result data (`Map<String, *>`)
- extensions (`Map<Object, Object>`)
- errors (`List<GraphQLError>`) 

So they are not available as strings like the input query is, I'll look into the object types as that might be okay, if not then I don't think this has to be graphQL specific necessarily and the approach with measuring at the request/response level would work better

Reply:
For a request body, one can just check the `Content-Length` header. For the response, I think it should be opt-in regardless due to the overhead of performing extra serialization.

See the related micrometer issue: https://github.com/micrometer-metrics/micrometer/issues/880

Reply:
What if we just add the metric for the request using the content-length as part of instrumentation for now. The response body metric is proving to be costly so I doubt it's a useful metric to have anyway.

Reply:
@kilink , yes, to get operationName for the metric tag, it seems like the request body would need to be deserialized in the filter. It would be cool if operationName was a expected header for graphql clients to provide, then, with the filter approach, getting the response size would be more efficient. Thanks for the suggestion of content length!
@srinivasankavitha yes, for now I will add the request size as that is a valuable metric to have and with content-length we can get the size efficiently in the instrumentation code, thanks for the feedback!